ddbToEsDlqLambda/index.ts

```diff
@@ -0,0 +1,102 @@
+import getComponentLogger from './loggerBuilder';
+
+const logger = getComponentLogger();
+const { QUEUE_URL, DDB_TO_ES_LAMBDA_NAME } = process.env;
+
+const sqs = new AWS.SQS();
+const lambda = new AWS.Lambda();
+const dynamodbstreams = new AWS.DynamoDBStreams();
+
+const getMessages = async (queueUrl) => {
+    const resp = await sqs
+        .receiveMessage({
+            QueueUrl: queueUrl,
+            MaxNumberOfMessages: 10,
+            VisibilityTimeout: 300,
+            WaitTimeSeconds: 0,
+        })
+        .promise();
+
+    return resp.Messages;
+};
+
+const deleteMessages = async (queueUrl, messagesToDelete) => {
+    await sqs
+        .deleteMessageBatch({
+            QueueUrl: queueUrl,
+            Entries: messagesToDelete,
+        })
+        .promise();
+};
+
+const invokeDdbToEsLambda = async (records) => {
+    await lambda
+        .invoke({
+            FunctionName: DDB_TO_ES_LAMBDA_NAME,
+            InvocationType: 'RequestResponse',
+            Payload: JSON.stringify(records),
+        })
+        .promise();
+};
+
+const getRecordsFromDbStream = async (message) => {
+    const resp = await dynamodbstreams
+        .getShardIterator({
+            ShardId: message.DDBStreamBatchInfo.shardId,
+            ShardIteratorType: 'AT_SEQUENCE_NUMBER',
+            StreamArn: message.DDBStreamBatchInfo.streamArn,
+            SequenceNumber: message.DDBStreamBatchInfo.startSequenceNumber,
+        })
+        .promise();
+
+    return dynamodbstreams
+        .getRecords({
+            ShardIterator: resp.ShardIterator,
+        })
+        .promise();
+};
+
+exports.handler = async () => {
+    let messages;
+    try {
+        messages = await getMessages(QUEUE_URL);
+        logger.debug('Recevied messages', JSON.stringify(messages));
+    } catch (e) {
+        logger.error('Failed to receive messages', JSON.stringify(e));
+    }
+
+    while (messages && messages.length > 0) {
+        const messagesToDelete = [];
+
+        /* eslint-disable no-restricted-syntax */
+        for (const message of messages) {
+            try {
+                const records = await getRecordsFromDbStream(JSON.parse(message.Body));
+                logger.debug('Fetched records', JSON.stringify(records));
+                await invokeDdbToEsLambda(records);
+                messagesToDelete.push({
+                    Id: message.MessageId,
+                    ReceiptHandle: message.ReceiptHandle,
+                });
+            } catch (e) {
+                logger.error('Failed to re-sync', JSON.stringify(e));
+            }
+        }
+
+        // only delete successfully re-synced messages
+        try {
+            await deleteMessages(QUEUE_URL, messagesToDelete);
+            logger.debug('Deleted messages', JSON.stringify(messagesToDelete));
+        } catch (e) {
+            logger.error('Failed to delete message', JSON.stringify(e));
+        }
+        try {
+            messages = await getMessages(QUEUE_URL);
```

Comment:
I feel like re-work the logic here to simplify it such that we don't need the extra call to getMessages() here. I just don't see it yet.

Reply:
I was thinking about this a little more, and I think there are several issues here:

1. There is the possibility for the function to run infinitely, or until Lambda kills it.
2. A SQS error probably should trigger the function to abort and return an error.

Here is psuedocode example of what I was thinking as to how we could re-work it.
```
try {
  // Process all SQS messages, or abort on error.
  while(true) {
	messages = getMessages(URL);
	if(messages.length <= 0) {
	  break;
	}
		
	for(message of messages) {
	  try {
	    ...
	  } catch(e) {
	    // DDBStream or Lambda error. Ignore for now, and continue processing messages.
	    log.error(..., e);
	    resync_error = true
	  }
	}
	
	if(msgs_to_delete.length > 0) {
	  // Delete all proccessed messages from the queue.
	}
  }
} catch(e) {
  // SQS error. Stop processing messages.
  log.error(..., e);
  sqs_error = true;
}

return !sqs_error || !resync_error;
```
We could introduce an upper limit on the # of messages that can be processed by each function call. The limit could be taken as input, or set as an environment variable, or both.
Here is what I was thinking as to how it would be used in the code above:
```
numMessagesToProcess = input.maxNumberOfMessages;
if(numMessagesToProcess == undefined) {
  numMessagesToProcess = MAX_NUMBER_OF_MESSAGES_TO_PROCESS;
}

try {
  // Process all SQS messages, or abort on error.
  while(numMessagesToProcess > 0) {
	messages = getMessages(URL, MESSAGE_BATCH_SIZE > numMessagesToProcess ? numMessagesToProcess : MESSAGE_BATCH_SIZE);
	if(messages.length <= 0) {
	  break;
	}
	numMessagesToProcess -= messages.length;

       ...
```

Reply:
This is good. I have one question about the final return `!sqs_error || !resync_error;` This would return false or true, but the lambda overall is always green.  Is that ok? Or we are expecting lambda is red when the returned value is false ? If we want the lambda red, we need to throw exceptions.

Reply:
Yeah, I guess I was expecting the Lambda function to fail if it returned a non-zero result, but if we need to throw an exception for that to happen, then so be it.

Reply:
FYI,  we can specify the max message we get from sqs queue. But the max number cannot be greater than 10.  I think there is no much value to specify batch size for method getMessages

Reply:
Actually even we specify max batch size as 10, in fact sqs returns less messages each time.

Reply:
On second thought, I think it is good to specify the batch size